pkg/nfd-master/nfd-master.go

```diff
@@ -432,6 +436,85 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// ensure we delete the old stale taints
+		labelReply, err := m.deleteTaints(c, r)
+		if err != nil {
+			return labelReply, err
+		}
+
+		if m.args.EnableTaints {
+			// Set NodeFeatureRule provided taints
+			if mustTaint {
+				labelReply, err := m.setTaints(c, r)
+				if err != nil {
+					return labelReply, err
+				}
+			}
+		}
+	}
+	return &pb.SetLabelsReply{}, nil
+}
+
+// setTaints sets taint(s) on the node matching rule(s) in the NodeFeatureRule CRD
+func (m *nfdMaster) setTaints(c context.Context, r *pb.SetLabelsRequest) (*pb.SetLabelsReply, error) {
+	var nodeTaints, crt []corev1.Taint
+	taintStrs := []string{}
+
+	err := authorizeClient(c, m.args.VerifyNodeName, r.NodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	klog.Infof("received tainting request for node %q", r.NodeName)
+
+	nfrs, err := m.nfdController.ruleLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to list NodeFeatureRule resources: %v", err)
+		return &pb.SetLabelsReply{}, err
+	}
+
+	cli, err := m.apihelper.GetClient()
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	// fetch the node object
+	node, err := m.apihelper.GetNode(cli, r.NodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	if !m.args.NoPublish {
```

Comment:
nit: I don't think we need to check this here as it's done already upper in the call chain (just makes the code a bit simpler)

Reply:
that's right, fixing it. Thanks